File: golang/internal/wow/formats/url_allowlist.go
package formats

import (
	"errors"
	"fmt"
	"net"
	"net/url"
	"strings"
)

var errFetchURLNotAllowed = errors.New("fetch URL host not allowed")

// ValidateFetchURL ensures url is HTTPS to an allowlisted public host (no private/metadata SSRF).
func ValidateFetchURL(raw string) error {
	parsed, err := url.Parse(raw)
	if err != nil {
		return fmt.Errorf("invalid fetch URL: %w", err)
	}
	if parsed.Scheme != "https" {
		return fmt.Errorf("%w: only https is allowed", errFetchURLNotAllowed)
	}
	if parsed.User != nil {
		return fmt.Errorf("%w: credentials in URL", errFetchURLNotAllowed)
	}
	host := parsed.Hostname()
	if host == "" {
		return fmt.Errorf("%w: missing host", errFetchURLNotAllowed)
	}
	if isDisallowedFetchHost(host) {
		return fmt.Errorf("%w: %s", errFetchURLNotAllowed, host)
	}
	if !isAllowedFetchHost(host) {
		return fmt.Errorf("%w: %s", errFetchURLNotAllowed, host)
	}
	return nil
}

func isDisallowedFetchHost(host string) bool {
	h := strings.ToLower(strings.TrimSuffix(host, "."))
	if h == "localhost" {
		return true
	}
	ip := net.ParseIP(h)
	if ip == nil {
		return false
	}
	return ip.IsLoopback() || ip.IsPrivate() || ip.IsLinkLocalUnicast() || ip.IsLinkLocalMulticast()
}

func isAllowedFetchHost(host string) bool {
	h := strings.ToLower(host)
	if isBlizzardCdnHost(h) {
		return true
	}
	if h == "github.com" || strings.HasSuffix(h, ".github.com") {
		return true
	}
	if strings.HasSuffix(h, ".githubusercontent.com") {
		return true
	}
	if h == "kruithne.net" || strings.HasSuffix(h, ".kruithne.net") {
		return true
	}
	return false
}

File: src/lib/wow/formats/url-allowlist.ts
const DISALLOWED_FETCH_HOSTS = new Set(['localhost']);

function isPrivateIP(host: string): boolean {
  const m = host.match(/^(\d+)\.(\d+)\.(\d+)\.(\d+)$/);
  if (!m) return false;
  const a = Number(m[1]);
  const b = Number(m[2]);
  if (a === 10) return true;
  if (a === 127) return true;
  if (a === 169 && b === 254) return true;
  if (a === 192 && b === 168) return true;
  if (a === 172 && b >= 16 && b <= 31) return true;
  return false;
}

function isBlizzardCdnHost(hostname: string): boolean {
  const h = hostname.toLowerCase();
  return /\.blizzard\.com$/i.test(h) || /\.battle\.net$/i.test(h);
}

function isAllowedFetchHost(hostname: string): boolean {
  const h = hostname.toLowerCase();
  if (isBlizzardCdnHost(h)) return true;
  if (h === 'github.com' || h.endsWith('.github.com')) return true;
  if (h.endsWith('.githubusercontent.com')) return true;
  if (h === 'kruithne.net' || h.endsWith('.kruithne.net')) return true;
  return false;
}

/** Ensures url is HTTPS to an allowlisted public host (blocks SSRF to metadata/private nets). */
export function validateFetchUrl(raw: string): URL {
  let parsed: URL;
  try {
    parsed = new URL(raw);
  } catch {
    throw new Error('Invalid fetch URL');
  }
  if (parsed.protocol !== 'https:') {
    throw new Error('Fetch URL must use HTTPS');
  }
  if (parsed.username || parsed.password) {
    throw new Error('Fetch URL must not include credentials');
  }
  const host = parsed.hostname.toLowerCase();
  if (!host || DISALLOWED_FETCH_HOSTS.has(host) || isPrivateIP(host) || !isAllowedFetchHost(host)) {
    throw new Error(`Fetch URL host not allowed: ${host}`);
  }
  return parsed;
}
